File: core/java/src/main/java/hara/lang/protocol/ICons.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.icons", name = "ICons")
public interface ICons<E> {
  @HaraMethod(value = "cons", arity = 2)
  ICons<E> cons(E e);
}

File: core/rust/src/lang/protocol/icons.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.icons", name = "ICons")]
pub trait ICons<E> {
    type Output;

    #[hara_method(value = "cons", arity = 2)]
    fn cons(&self, value: E) -> Self::Output;
}
